type_decoders/README.md

```diff
@@ -0,0 +1,787 @@
+```
+
+`isInstanceOf()` accepts a javascript constructor argument and creates a decoder which verifies that its input is `instanceof clazz`.
+
+### isCheckedWith()
+
+```ts
+interface ICheckedWithDecoderOptions {
+  msg?: DecoderErrorMsgArg;
+  promise?: boolean;
+}
+
+function isCheckedWith<T>(fn: (value: T) => boolean, options?: ICheckedWithDecoderOptions): Decoder<T, T>;
+function isCheckedWith<T>(fn: (value: T) => Promise<boolean>, options: ICheckedWithDecoderOptions & { promise: true }): PromiseDecoder<T, T>;
+```
+
+`isCheckedWith()` accepts a predicate function argument and creates a decoder which verifies that inputs pass the function check.
+
+**Async**: to pass a predicate function which returns a promise resolving to a boolean, pass the `promise: true` option to `isCheckedWith()`.
+
+### isOptional()
+
+```ts
+interface IOptionalDecoderOptions {
+  msg?: DecoderErrorMsgArg;
+}
+
+function isOptional<T>(decoder: Decoder<T>, options?: IOptionalDecoderOptions): Decoder<T | undefined>;
+function isOptional<T>(decoder: PromiseDecoder<T>, options?: IOptionalDecoderOptions): PromiseDecoder<T | undefined>;
+```
+
+`isOptional()` accepts a decoder and returns a new decoder which accepts either the original decoder's value or `undefined`. Convenience function for `isAnyOf([decoder, isUndefined()])`.
+
+### isNullable()
+
+```ts
+interface INullableDecoderOptions {
+  msg?: DecoderErrorMsgArg;
+}
+
+function isNullable<T>(decoder: Decoder<T>, options?: INullableDecoderOptions): Decoder<T | null>;
+function isNullable<T>(decoder: PromiseDecoder<T>, options?: INullableDecoderOptions): PromiseDecoder<T | null>;
+```
+
+`isNullable()` accepts a decoder and returns a new decoder which accepts either the original decoder's value or `null`. Convenience function for `isAnyOf([decoder, isNull()])`.
+
+### isMaybe()
+
+```ts
+interface IMaybeDecoderOptions {
+  msg?: DecoderErrorMsgArg;
+}
+
+function isMaybe<T>(decoder: Decoder<T>, options?: IMaybeDecoderOptions): Decoder<T | null | undefined>;
+function isMaybe<T>(decoder: PromiseDecoder<T>, options?: IMaybeDecoderOptions): PromiseDecoder<T | null | undefined>;
+```
+
+`isMaybe()` accepts a decoder and returns a new decoder which accepts either the original decoder's value or `null` or `undefined`. Convenience function for `isAnyOf([decoder, isNull(), isUndefined()])`.
+
+### isAnyOf()
+
+```ts
+interface IAnyOfDecoderOptions {
+  msg?: DecoderErrorMsgArg;
+}
+
+function isAnyOf<T extends Decoder<unknown>>(decoders: T[], options?: IAnyOfDecoderOptions): Decoder<DecoderReturnType<T>>;
+function isAnyOf<T extends Decoder<unknown> | PromiseDecoder<unknown>>(decoders: T[], options?: IAnyOfDecoderOptions): PromiseDecoder<DecoderReturnType<T>>;
+```
+
+`isAnyOf()` accepts an array of decoders and attempts to decode a provided value using each of them, in order, returning the first successful result or `DecoderError` if all fail.
```

Comment:
Which `DecoderError` would be returned here? Would it create its own or somehow combine the errors of the passed decoders?

Reply:
`isAnyOf` is unique in that it doesn't return the nested `DecoderErrors`, and instead returns it's own error. By default, the error message is `(child error messages[]).join(' OR ')`. A current limitation of this module's API is that only a single `DecoderError` (the first one) is returned (i.e. you can't collect all the errors associated with an input). Because of this, it isn't currently possible to return multiple nested errors with a isAnyOf` DecoderError. I'm pretty convinced that the final API should support multiple nested errors though. Though this shouldn't be a large change.

Reply:
It’s also a little difficult to convey since you’d normally expect at least some of the decoders to fail even if the `anyOf` succeeds.